Use case demo (#835)
EventStore/documentation: docs/getting-started/use-cases/mix-and-match-database/explanation.md

---
title: Explanation
---

# Explanation

![](./images/use-the-right-database-for-the-job.png)

## Mix-and-Match Database

Modern applications often deal with diverse data access patterns that a single database can’t efficiently handle. That’s where you can leverage a database mix-and-match approach where different types of databases for different parts of your system can be used based on their strengths.

For example:

- **PostgreSQL** for general queries.
- **Redis** for fast lookups.
- **MongoDB** for web frontends.

This strategy is beneficial because no single database is perfect for every use case. Selecting the most optimal database for each specific task enhances performance, scalability, and developer efficiency.

::: info

This is also popularly known as **Polyglot Persistence**. 

:::

## Traditional Approach

> Traditionally, systems use a single relational database for all operations and it is not simple or reliable to push updates to other data stores.

Reviewer: ```suggestion
Traditionally, systems use a single relational database for all operations, and pushing updates to other data stores is not reliable or straightforward.
```

To distribute data across multiple databases, developers typically implement synchronization mechanisms like:

- **Manual synchronization and ETL jobs** to periodically replicate data.
- **Message brokers** (e.g., Kafka, RabbitMQ) that propagate updates to various services.

However, these methods can have drawbacks:

- Risk of synchronization failure.
- Noticeable delays due to batch processing.
- Difficulty in recovery after failures.
- Issues with message loss, duplication, or ordering.

Such complexities increase the developers' workload to maintain synchronization and consistency.

> ## Mix-and-Match Database Effectively with KurrentDB

Reviewer: ```suggestion
## Mix-and-Match Databases Effectively with KurrentDB
```

> KurrentDB captures all changes (events) immutablly within your system, which can be consumed by any number of downstream systems or databases in real time. This means you can reliably synchronize a variety of databases or data models, each optimized for a specific use case, without the complexity, risks, or delays of traditional batch jobs or external messaging systems.

Reviewer: ```suggestion
KurrentDB captures all changes (events) immutably within your system, which can be consumed by any number of downstream systems or databases in real time. This means you can reliably synchronize a variety of databases or data models, each optimized for a specific use case, without the complexity, risks, or delays of traditional batch jobs or external messaging systems.
```
